File: Data Structures/Trees/Depth-First Search/Implementation.go
package main

import "fmt"

// TreeNode represents a node in a binary tree
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// NewTreeNode creates a new TreeNode with the given value
func NewTreeNode(val int) *TreeNode {
	return &TreeNode{
		Val:   val,
		Left:  nil,
		Right: nil,
	}
}

// DFS represents a depth-first search object
type DFS struct {}

// InOrder performs an in-order traversal of a binary tree
func (dfs *DFS) InOrder(root *TreeNode) {
	if root == nil {
		return
	}
	dfs.InOrder(root.Left) // Traverse left subtree
	fmt.Print(root.Val)   // Print the current node's value
	dfs.InOrder(root.Right) // Traverse right subtree
}

// PreOrder performs a pre-order traversal of a binary tree
func (dfs *DFS) PreOrder(root *TreeNode) {
	if root == nil {
		return
	}
	fmt.Print(root.Val)   // Print the current node's value
	dfs.PreOrder(root.Left) // Traverse left subtree
	dfs.PreOrder(root.Right) // Traverse right subtree
}

// PostOrder performs a post-order traversal of a binary tree
func (dfs *DFS) PostOrder(root *TreeNode) {
	if root == nil {
		return
	}
	dfs.PostOrder(root.Left) // Traverse left subtree
	dfs.PostOrder(root.Right) // Traverse right subtree
	fmt.Print(root.Val)   // Print the current node's value
}

// time complexity: O(n)
// space complexity: O(n)

File: Data Structures/Trees/Depth-First Search/Implementation.py
class TreeNode:
    def __init__(self, val):
        """
        Initialize a new TreeNode with the given value.

        Args:
            val (int): The value to be assigned to the node.

        Returns:
            None
        """
        self.val = val
        self.left = None
        self.right = None

def inorder(root):
    """
    Perform an inorder traversal of a binary tree.

    Args:
        root (TreeNode): The root node of the binary tree.

    Returns:
        None
    """
    if not root:
        return
    inorder(root.left)  # Traverse left subtree
    print(root.val)     # Print value of current node
    inorder(root.right) # Traverse right subtree

def preorder(root):
    """
    Perform a preorder traversal of a binary tree.

    Args:
        root (TreeNode): The root node of the binary tree.

    Returns:
        None
    """
    if not root:
        return
    print(root.val)     # Print value of current node
    preorder(root.left) # Traverse left subtree
    preorder(root.right) # Traverse right subtree

def postorder(root):
    """
    Perform a postorder traversal of a binary tree.

    Args:
        root (TreeNode): The root node of the binary tree.

    Returns:
        None
    """
    if not root:
        return
    postorder(root.left)  # Traverse left subtree
    postorder(root.right) # Traverse right subtree
    print(root.val)       # Print value of current node

# time complexity: O(n)
# space complexity: O(n)